integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/TableContentChangeDatasetBuilder.java

```diff
@@ -83,11 +84,36 @@ protected List<OpenLineage.OutputDataset> apply(SparkListenerEvent event, Logica
                   OpenLineage.LifecycleStateChangeDatasetFacet.LifecycleStateChange.OVERWRITE,
                   null));
     }
+
+    // FIXME: Use 'castToDataSourceV2Relation()' to safely cast the relation to
+    // 'DataSourceV2Relation'. We are unsure of the plan structure that would cause a
+    // 'ClassCastException' to be thrown; therefore, to get meaningful insight we also log the
+    // logical plan when the relation is of the type 'DataSourceV2ScanRelation'.
+    final DataSourceV2Relation returnTable = castToDataSourceV2Relation(x, table);
     if (includeDatasetVersion(event)) {
       DatasetVersionDatasetFacetUtils.includeDatasetVersion(
-          context, datasetFacetsBuilder, (DataSourceV2Relation) table);
+          context, datasetFacetsBuilder, returnTable);
     }
+
     return PlanUtils3.fromDataSourceV2Relation(
-        outputDataset(), context, (DataSourceV2Relation) table, datasetFacetsBuilder);
+        outputDataset(), context, returnTable, datasetFacetsBuilder);
+  }
+
+  private DataSourceV2Relation castToDataSourceV2Relation(LogicalPlan x, NamedRelation table) {
+    log.debug("Attempting to cast '{}' to 'DataSourceV2Relation'...", table.getClass());
+    // Ensure the relation is not a scan relation.
+    if (table instanceof DataSourceV2ScanRelation) {
```

Comment:
Minor point, but I'd do the `if instanceof` check prior to calling this method - there's no point in the extra method call for like 99.9% of cases

Reply:
That's a fair point. I'll adjust the placing of `instanceof`.